### PR TITLE
Use GMPY factorial function for integers

### LINKED ISSUE #2704: faster factorials (and other functions) using gmpy
Sympy's current factorial function sympy.functions.combinatorial.factorials.factorial(x) is much slower than using gmpy, then converting to a sympy integer: S(gmpy.fac(x)). Since we already take advantage of gmpy for ground types, doesn't it make sense to do the same for integer functions, where possible (and where it helps)?

Another example of where this makes a big difference is in computing Fibonacci numbers. For example at 2**18, sympy allocates at least a gig of memory, while gmpy allocates none and returns instantly. I killed the sympy calculation pretty soon after I started it, so I don't know how long it would take to run to completion, but it is obviously slower than gmpy and the memory issue alone would make gmpy an improvement. 

sympy's binomial function actually computes C(2**19, 2**18) faster than gmpy. Curiously, gmpy.bincoef is also slower than using gmpy's own factorial function and the definition of the binomial coefficient.

I have some benchmark numbers on factorials (and a scripti to generate them, plus a plot) if anyone is interested. As an example, sympy takes 8.33s to compute (2**20)!, whereas S(gmpy.fac(2**20)) takes 0.41s (a 20x speedup). I see similar speed-ups for numbers in the range of 2**10 to 2**21. Memory usage is not an issue here, thankfully.

### PR DESCRIPTION
**Pull request by langston-barrett**
Should be faster to use C than Python for this. See the issue referenced for some benchmarks. 

Fixes #2704

### CODE REVIEW

**Comment by siddharthist** [comment]
The test failures don't look related to my changes, should I do anything about that?

**Comment by isuruf** [comment]
+1. Can you fix the conflicts?

**Comment by siddharthist** [comment]
@isuruf Fixed!

File: sympy/functions/combinatorial/factorials.py [hunk 1]
```diff
@@ -162,6 +167,13 @@ def eval(cls, n):
                                 cls._small_factorials.append(result)
                         result = cls._small_factorials[n-1]
 
+                    # GMPY factorial is faster, use it when available
+                    elif HAS_GMPY == 1:
+                        result = gmpy_factorial(n)
+
+                    elif HAS_GMPY == 2:
+                        result = gmpy2_factorial(n)
+
```
Comment (isuruf) [inline sympy/functions/combinatorial/factorials.py]: You can do,
```
elif HAS_GMPY:
    from sympy.core.compatibility import gmpy
    result = gmpy.fac(n)
```
This should work for `gmpy` and `gmpy2`

**Comment by siddharthist** [comment]
@isuruf Changed

**Comment by isuruf** [comment]
+1 to merge. This is faster when gmpy is available.

**Comment by siddharthist** [comment]
@isuruf Any update?

**Comment by isuruf** [comment]
@asmeurer, ping

**Pull request merged**